backgroundremover: init at 0.2.1

JayRovacsek opened the pull request:
###### Description of changes

This change set implements the backgroundremover python package and requisite dependencies that do not currently exist within nixpkgs.
Backgroundremover is a package to remove backgrounds from images and video and the homepage can be found [here](https://github.com/nadermx/backgroundremover)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

edit: Please note this pull request supersedes https://github.com/NixOS/nixpkgs/pull/231295

Closes https://github.com/NixOS/nixpkgs/issues/230522

JayRovacsek committed:
maintainers: add jayrovacsek

JayRovacsek committed:
commandlines: init at 0.4.1

JayRovacsek committed:
hsh: init at 1.1.0

JayRovacsek committed:
PyMatting: init at 1.1.8

JayRovacsek committed:
backgroundremover: init at 0.2.1

JayRovacsek commented:
@lucasew @JulienMalka I'm sorry to tag you both here, just resolving issues presented in https://github.com/NixOS/nixpkgs/pull/231295 - I will be posting results of nixpkgs-review shortly including your recommendations of the `--post-result` option (thanks Lucas!) for both x86 and aarch linux
Further to this, will have access to an aarch darwin machine today so will be able to build and post results for those systems also

JayRovacsek commented:
Noting again, closes: https://github.com/NixOS/nixpkgs/issues/230522

lucasew commented:
> Noting again, closes: #230522

For the automation to work you must put "Closes #something" in the description. When this get merged those issues are closed automatically

JayRovacsek commented:
Agh! In migrating changes over I've stuffed up slightly. The python-modules folder has incorrectly been placed in an incorrect scope :facepalm: 

Changes incoming...

lucasew commented:
> Agh! In migrating changes over I've stuffed up slightly. The python-modules folder has incorrectly been placed in an incorrect scope facepalm
> 
> Changes incoming...

Yeah, bruh

```
error:
       … while evaluating call site

       at «none»:0: (source not available)

       … while calling 'func'

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/pkgs/development/interpreters/python/default.nix:40:24:

           39|           valid = value: pythonPackages.hasPythonModule value || providesSetupHook value || lib.elem value exceptions;
           40|           func = name: value:
             |                        ^
           41|             if lib.isDerivation value then

       … while evaluating call site

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/pkgs/development/interpreters/python/default.nix:41:16:

           40|           func = name: value:
           41|             if lib.isDerivation value then
             |                ^
           42|               lib.extendDerivation (valid value || throw "${name} should use `buildPythonPackage` or `toPythonModule` if it is to be part of the Python packages set.") {} value

       … while calling 'isDerivation'

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/lib/attrsets.nix:636:5:

          635|     # Value to check.
          636|     value: value.type or null == "derivation";
             |     ^
          637|

       … while evaluating call site

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/pkgs/top-level/python-packages.nix:1146:23:

         1145|
         1146|   backgroundremover = callPackage ../development/python-modules/backgroundremover { };
             |                       ^
         1147|

       … while calling 'callPackageWith'

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/lib/customisation.nix:128:35:

          127|   */
          128|   callPackageWith = autoArgs: fn: args:
             |                                   ^
          129|     let

       … while evaluating call site

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/lib/customisation.nix:141:10:

          140|         # Filter out arguments that have a default value
          141|         (lib.filterAttrs (name: value: ! value)
             |          ^
          142|         # Filter out arguments that would be passed

       … while calling 'filterAttrs'

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/lib/attrsets.nix:305:5:

          304|     # The attribute set to filter
          305|     set:
             |     ^
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));

       … while evaluating call site

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/lib/customisation.nix:131:15:

          130|       f = if lib.isFunction fn then fn else import fn;
          131|       fargs = lib.functionArgs f;
             |               ^
          132|

       … while calling 'functionArgs'

       at /home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/lib/trivial.nix:440:18:

          439|   */
          440|   functionArgs = f:
             |                  ^
          441|     if f ? __functor

       error: getting status of '/home/lucasew/.cache/nixpkgs-review/pr-231950/nixpkgs/pkgs/development/python-modules/backgroundremover': No such file or directory
https://github.com/NixOS/nixpkgs/pull/231950 failed to build
$ git worktree prune
```

JayRovacsek commented:
Result of `nixpkgs-review pr 231950` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

JayRovacsek commented:
Result of `nixpkgs-review pr 231950` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

JayRovacsek commented:
Result of `nixpkgs-review pr 231950` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

JayRovacsek commented:
Result of `nixpkgs-review pr 231950` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

lucasew commented:
Result of `nixpkgs-review pr 231950` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

JayRovacsek commented:
Thanks very much for your help and feedback @lucasew!

alyssais commented:
Something has gone wrong with your commits here.  You have one set of commits adding packages into pkgs/development/python-modules (correct), and then another set of commits copying the same files into pkgs/python-modules.

JayRovacsek commented:
Rebased, removed odd commits and force pushed, nice pickup @alyssais! I believe this was an error on my part attempting to rewrite some of the commits once I'd already created the pull :grimacing: 
Happy to run nixpkgs-review again if it's preferred/required :+1:

lucasew commented:
Result of `nixpkgs-review pr 231950` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

JayRovacsek commented on pkgs/development/python-modules/backgroundremover/default.nix:
```diff
@@ -0,0 +1,77 @@
+
+  meta = with lib; {
+    description =
+      "BackgroundRemover is a command line tool to remove background from image and video, made by nadermx to power https://BackgroundRemoverAI.com";
```
Ah thanks @alyssais ! So yes it could be utilised as both a library as well as a binary, would you prefer I move it into the correct location (top-level/all-packages.nix) as the initial commit or is it okay to exist as a new commit?
Further to this, I note `hsh` falls into the same category so I'll resolve for that also

alyssais replied:
I think you have it the wrong way round — if it's both, keep it as is, but add a `toPythonApplication` line in all-packages.nix.  Please make the changes in the commits that add each package.

JayRovacsek committed:
Merge branch 'master' into backgroundremover

JayRovacsek commented:
Closing in favour of https://github.com/NixOS/nixpkgs/pull/277394 - thanks @lucasew
